docs/incubator-proposal.md

```diff
@@ -0,0 +1,188 @@
+    # configured in the letsencrypt-staging issuer above.
+    - domains:
+      - cm-dns-clouddns.k8s.group
+      dns-01:
+        provider: clouddns-prod
+```
+
+These example manifests do not describe a finalised API, but instead aim to
+help communicate the concept of Issuer vs Certificate.
+
+## Goals
+
+* Provide a non-opinionated interface for new issuers to be added to
+cert-manager
+
+* Provide reliable support for at least ACME HTTP01 and DNS01 challenge types
+in the ACME provider. TLS-SNI-02 support is not a goal listed here due to
+potential difficulties in implementation with the current Ingress controller
+specification. In future, TLS-SNI-02 support for ACME may be considered.
+
+* Provide clear, understable Events/log messages to users to inform them that
+issuance/renewal of a certificate has failed. Also provide metrics endpoints
+that can be scrapped by metrics collection software (such as Prometheus).
+
+* Attempt to obtain certificates from the specified issuing backend, and then
+keep those certificates up to date by renewing them as they near expiry.
+
+* Provide a relatively easy upgrade path for existing users of kube-lego and
+kube-cert-manager. Whilst we won't have 1:1 compatility, the concepts should
+be relatively similar/transfer across well.
+
+## Non-goals
+
+* We specifically do not want to tie ourselves to Ingress resources, as
+previous projects in this space have done. With the out-of-tree nature of
+Ingress, and with some areas having loose specification (ie.
+[kubernetes/ingress#555](https://github.com/kubernetes/ingress/issues/555),
+it's desired that we don't specifically list a set of 'supported' Ingress
+controllers. This particularly limitation specifically applies to the ACME
+challenge provider at this point.
+
+* Tying ourselves to any specific issuer implementation. Whilst previous work
+has focused heavily on ACME, we do not want to be in any way tied to this
+provider. Additional issuers such as [Hashicorp Vault](https://vaultproject.io)
+are planned in future.
+
+## Roadmap
+
+### Short term
+
+* Stabilise application to bring it to a comparable state to kube-lego and
+kube-cert-manager.
+
+* Implement a fairly robust test suite so we can confidently move forwards.
+
+* Creating our own Controller and resource types, namely 'Issuer' and
+'Certificate'.
+
+* Writing up retrospective design proposals for what we have already.
+
+* Filling out documentation and creating tutorials to make the project
+user-friendly from day one. Tutorials should be clear and cover common
+user stories.
+
+### Medium term
+
+* Implement a metrics endpoint to expose useful information such as number
+of certificates issued, number of attempts per certificate, etc.
+
+* Implement API server that registers with aggregator, or utilise new
+validation APIs in future Kubernetes versions.
+
+### Longer term
+
+* Implementing additional certificate issuers (eg. Vault, Amazon?).
```

Comment:
I realise having Vault issue certs is on the long term plan, but I was wondering if you have a timeframe in which you might get to those goals? Being able to issue certs from Vault, without having to use LetsEncrypt, is the main reason I'm really keen on the work you're doing on this project.

Reply:
FYI https://github.com/giantswarm/cert-operator is providing certificate generation using the PKI backend. We thought about contributing here at some point or at least try to align the interface (TPR/CRD) structure for some more common usage so you can plug and play out of the box. That effort was just not yet done.